File: app/src/main/java/com/coinninja/coinkeeper/ui/dropbit/me/DropbitMeConfiguration.java
package com.coinninja.coinkeeper.ui.dropbit.me;

import android.net.Uri;

import com.coinninja.coinkeeper.di.interfaces.CoinkeeperApplicationScope;
import com.coinninja.coinkeeper.di.interfaces.DropbitMeUri;
import com.coinninja.coinkeeper.model.db.Account;
import com.coinninja.coinkeeper.model.db.DropbitMeIdentity;
import com.coinninja.coinkeeper.model.helpers.WalletHelper;

import java.util.List;

import javax.inject.Inject;

@CoinkeeperApplicationScope
public class DropbitMeConfiguration {

    private final Uri dropbitMeUri;
    private final WalletHelper walletHelper;
    private boolean isNewlyVerified = false;
    private boolean shouldPromptWhenNextAvailable = false;
    private OnViewDropBitMeViewRequestedObserver onViewDropBitMeViewRequestedObserver;

    @Inject
    DropbitMeConfiguration(@DropbitMeUri Uri dropbitMeUri, WalletHelper walletHelper) {
        this.dropbitMeUri = dropbitMeUri;
        this.walletHelper = walletHelper;
    }

    public void setNewlyVerified() {
        isNewlyVerified = true;
        shouldPromptWhenNextAvailable = true;
    }

    public boolean isNewlyVerified() {
        return isNewlyVerified;
    }

    public boolean shouldShowWhenPossible() {
        return shouldPromptWhenNextAvailable;
    }

    public void showWhenPossible() {
        shouldPromptWhenNextAvailable = true;

        if (onViewDropBitMeViewRequestedObserver != null) {
            onViewDropBitMeViewRequestedObserver.onShowDropBitMeRequested();
        }
    }

    public void acknowledge() {
        isNewlyVerified = false;
        shouldPromptWhenNextAvailable = false;
    }

    public boolean hasVerifiedAccount() {
        return walletHelper.hasVerifiedAccount();
    }


    public void setOnViewDropBitMeViewRequestedObserver(OnViewDropBitMeViewRequestedObserver onViewDropBitMeViewRequestedObserver) {
        this.onViewDropBitMeViewRequestedObserver = onViewDropBitMeViewRequestedObserver;
    }

    public String getShareUrl() {
        Account userAccount = walletHelper.getUserAccount();
        if (userAccount == null || userAccount.getIdentities().size() == 0) return "";

        List<DropbitMeIdentity> identities = userAccount.getIdentities();
        return dropbitMeUri.buildUpon().appendPath(identities.get(0).getHandle()).toString();
    }

    public boolean isDisabled() {
        Account userAccount = walletHelper.getUserAccount();
        return userAccount == null || userAccount.getIsPrivate();
    }
}

File: app/src/main/java/com/coinninja/coinkeeper/ui/dropbit/me/DropbitMeConfiguration.kt
package com.coinninja.coinkeeper.ui.dropbit.me

import android.net.Uri
import app.dropbit.annotations.Mockable
import com.coinninja.coinkeeper.di.interfaces.CoinkeeperApplicationScope
import com.coinninja.coinkeeper.di.interfaces.DropbitMeUri
import com.coinninja.coinkeeper.model.db.DropbitMeIdentity
import com.coinninja.coinkeeper.model.helpers.DropbitAccountHelper
import com.coinninja.coinkeeper.twitter.MyTwitterProfile
import java.net.URLDecoder
import javax.inject.Inject

@Mockable
@CoinkeeperApplicationScope
class DropbitMeConfiguration @Inject
internal constructor(@DropbitMeUri internal val dropbitMeUri: Uri,
                     internal val dropbitAccountHelper: DropbitAccountHelper,
                     val myTwitterProfile: MyTwitterProfile) {

    var isNewlyVerified = false
        internal set

    internal var shouldPromptWhenNextAvailable = false
    internal var onViewDropBitMeViewRequestedObserver: OnViewDropBitMeViewRequestedObserver? = null

    val shareUrl: String
        get() {
            val identity: DropbitMeIdentity? = dropbitAccountHelper.preferredIdentity

            identity?.let { identity ->
                val handle = identity.handle
                return URLDecoder.decode(dropbitMeUri.buildUpon().appendPath(handle).toString(), "UTF-8")
            }
            return ""
        }

    val isDisabled: Boolean
        get() = dropbitAccountHelper.hasPrivateAccount

    fun setIsNewlyVerified() {
        isNewlyVerified = true
    }

    fun getAvatar(): String? {
        return myTwitterProfile.myUser?.profileImage
    }

    fun setInitialVerification() {
        isNewlyVerified = true
        shouldPromptWhenNextAvailable = true
    }

    fun shouldShowWhenPossible(): Boolean {
        return shouldPromptWhenNextAvailable
    }

    fun showWhenPossible() {
        shouldPromptWhenNextAvailable = true
        onViewDropBitMeViewRequestedObserver?.onShowDropBitMeRequested()
    }

    fun acknowledge() {
        isNewlyVerified = false
        shouldPromptWhenNextAvailable = false
    }

    fun hasVerifiedAccount(): Boolean = dropbitAccountHelper.hasVerifiedAccount

    fun setOnViewDropBitMeViewRequestedObserver(onViewDropBitMeViewRequestedObserver: OnViewDropBitMeViewRequestedObserver?) {
        this.onViewDropBitMeViewRequestedObserver = onViewDropBitMeViewRequestedObserver
    }
}
